src/_pytest/config/__init__.py

```diff
@@ -1308,15 +1323,22 @@ def parse(self, args: List[str], addopts: bool = True) -> None:
         self.hook.pytest_cmdline_preparse(config=self, args=args)
         self._parser.after_preparse = True  # type: ignore
         try:
+            source = Config.ArgsSource.ARGS
             args = self._parser.parse_setoption(
                 args, self.option, namespace=self.option
             )
             if not args:
                 if self.invocation_params.dir == self.rootpath:
-                    args = self.getini("testpaths")
+                    args = []
+                    source = Config.ArgsSource.TESTPATHS
+                    testpaths = cast(List[str], self.getini("testpaths"))
```

Comment:
Is a `cast` really needed? Doesn't

```suggestion
                    testpaths: List[str] = self.getini("testpaths")
```

(like done in `terminal.py` too) work as well?